### PR TITLE
CP-26627 Isolate sriov/sriov vlan network from other network

### PR DESCRIPTION
**Pull request by krizex**
1. when creating vlan, check the candidate pif type and the first type of the first
existing pif of the network. Only allow the creation if:
  a. both candidate PIF and existing PIF are sriov logical PIF and share same
  PCI device type
  or
  b. both candidate PIF and existing PIF are not sriov logical PIF

2. Add UTs accordingly

Signed-off-by: Yang Qian <yang.qian@citrix.com>

---
This change is [Reviewable](https://reviewable.io/reviews/xapi-project/xen-api/3479)

### CODE REVIEW

**Comment by mseri** [comment]
Not for the next time: this PR should have had at least two separate commits, one introducing the tests and one with the code changes.

File: ocaml/idl/datamodel.ml [hunk 1]
```diff
@@ -571,6 +571,8 @@ let _ =
     ~doc:"Failed to disable SR-IOV on PIF" ();
   error Api_errors.network_has_incompatible_sriov_pifs ["PIF"; "network"]
     ~doc:"The PIF is not compatible with the selected SR-IOV network" ();
+  error Api_errors.vlan_network_incompatible ["PIF"; "network"]
```
Comment (mseri) [inline ocaml/idl/datamodel.ml]: Below we call the error `network_incompatible_purposes`, why not using a similar same style here as well and call this `network_incompatible_with_vlan`. Or `cannot_add_vlan_to_network` like the errors further down. We are already inconsistent enough, I think we should try to stick to some styles

File: ocaml/xapi-consts/api_errors.ml [hunk 2]
```diff
@@ -100,6 +100,7 @@ let network_sriov_enable_failed = "NETWORK_SRIOV_ENABLE_FAILED"
 let network_sriov_disable_failed = "NETWORK_SRIOV_DISABLE_FAILED"
 let network_is_not_sriov_compatible = "NETWORK_IS_NOT_SRIOV_COMPATIBLE"
 let network_has_incompatible_sriov_pifs = "NETWORK_HAS_INCOMPATIBLE_SRIOV_PIFS"
+let vlan_network_incompatible = "VLAN_NETWORK_INCOMPATIBLE"
```
Comment (mseri) [inline ocaml/xapi-consts/api_errors.ml]: Update this depending on how you update the error in `datamodel.ml`

File: ocaml/xapi/xapi_vlan.ml [hunk 3]
```diff
@@ -43,13 +43,38 @@ let create_internal ~__context ~host ~tagged_PIF ~tag ~network ~device =
   let () = Db.VLAN.create ~__context ~ref:vlan ~uuid:vlan_uuid ~tagged_PIF ~untagged_PIF ~tag ~other_config:[] in
   vlan, untagged_PIF
 
+let assert_vlan_on_pif_compatible_with_network ~__context candidate_pif candidate_pif_topo network =
```
Comment (mseri) [inline ocaml/xapi/xapi_vlan.ml]: Use named arguments:
```ocaml
let assert_vlan_on_pif_compatible_with_network ~__context ~candidate_pif ~candidate_pif_topo ~network = [...]
```

File: ocaml/xapi/xapi_vlan.ml [hunk 4]
```diff
@@ -43,13 +43,38 @@ let create_internal ~__context ~host ~tagged_PIF ~tag ~network ~device =
...
+  let open Xapi_pif_helpers in
+  match Db.Network.get_PIFs ~__context ~self:network with
+  | existing_pif :: _ ->
```
Comment (mseri) [inline ocaml/xapi/xapi_vlan.ml]: Simplify the code (also double check if you still need the `begin`/`end`, I think not):
```ocaml
match Db.Network.get_PIFs ~__context ~self:network with
| [] -> ()
| pif :: _ ->
    let pif_rec = Db.PIF.get_record ~__context ~self:pif in
    let pif_topo = get_pif_topo ~__context ~pif_rec in
    match pif_topo, candidate_pif_topo with
    | VLAN_untagged _ :: Network_sriov_logical existing_sriov :: _
    , Network_sriov_logical candidate_sriov :: _ ->
        (* both based on sriov logical, then compare sriov compatibility *)
        let existing = Db.Network_sriov.get_physical_PIF ~__context ~self:existing_sriov in
        let candidate = Db.Network_sriov.get_physical_PIF ~__context ~self:candidate_sriov in
        if not (is_device_underneath_same_type ~__context existing candidate) then
          raise Api_errors.(Server_error (vlan_network_incompatible, [Ref.string_of candidate; Ref.string_of network]))
    | VLAN_untagged _ :: Network_sriov_logical _ :: _, _
        (* existing pif based on sriov logical but candidate pif does not *)
    | VLAN_untagged _ :: _, Network_sriov_logical _ :: _ ->
        (* existing pif not based on sriov logical but candidate pif does *)
        raise Api_errors.(Server_error (vlan_network_incompatible, [Ref.string_of candidate; Ref.string_of network]))
    | _ -> ()
```

**Comment by robhoes** [comment]
Only when creating VLANs? I thought that the plan was to disallow mixing SR-IOV PIFs with _any_ other PIFs on a network? Then you should should assert that for all of `bond.create`, `vlan.create` and `tunnel.create`, the given network has no SR-IOV PIFs, and that for `network_sriov.create`, the given network has not non-SR-IOV PIFs.

File: ocaml/idl/datamodel.ml (hunk 1 above)
Comment (krizex) [reply ocaml/idl/datamodel.ml]: Yes, I think `network_incompatible_with_X` is better.

File: ocaml/xapi-consts/api_errors.ml (hunk 2 above)
Comment (krizex) [reply ocaml/xapi-consts/api_errors.ml]: Done

File: ocaml/xapi/xapi_vlan.ml (hunk 3 above)
Comment (krizex) [reply ocaml/xapi/xapi_vlan.ml]: Done

File: ocaml/xapi/xapi_vlan.ml (hunk 4 above)
Comment (krizex) [reply ocaml/xapi/xapi_vlan.ml]: I cannot find out how to simply the code anymore except remove the `begin` and `end`.

**Comment by krizex** [comment]
@mseri Good suggestion, I will make that change and also port the review of this PR to reviewable to help you review the code, thank you!

**Comment by krizex** [comment]
@robhoes Yes, you are right, I missed them.

**Comment by krizex** [comment]
@mseri @robh Updated, please help review again, thank you.

**Comment by robhoes** [comment]
Reviewed 2 of 2 files at r1, 4 of 4 files at r2, 7 of 7 files at r3.
Review status: 9 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[ocaml/xapi/xapi_network_helpers.ml, line 41 at r3](https://reviewable.io:443/reviews/xapi-project/xen-api/3479#-L6R3iybjTpWd2ei4T5D:-L6R3iybjTpWd2ei4T5E:b-9olk57) ([raw file](https://github.com/xapi-project/xen-api/blob/d84ad666a28e89efffa83fd40e0bef9ee64caffb/ocaml/xapi/xapi_network_helpers.ml#L41)):*
> ```OCaml
>     raise Api_errors.(Server_error (network_incompatible_with_vlan_on_bridge, [Ref.string_of network]))
> 
> let assert_network_compatible_with_vlan_on_sriov ~__context ~network ~sriov ~tagged_PIF =
> ```

There is another related function in xapi_network_sriov_helpers.ml: `assert_sriov_pif_compatible_with_network`.

---

*[ocaml/xapi/xapi_network_helpers.ml, line 48 at r3](https://reviewable.io:443/reviews/xapi-project/xen-api/3479#-L6R2dgwSBwZrZmaCQfw:-L6R2dgwSBwZrZmaCQfx:boytzbh) ([raw file](https://github.com/xapi-project/xen-api/blob/d84ad666a28e89efffa83fd40e0bef9ee64caffb/ocaml/xapi/xapi_network_helpers.ml#L48)):*
> ```OCaml
>     match get_pif_topo ~__context ~pif_rec:existing_pif_rec with
>     | VLAN_untagged _ :: Network_sriov_logical existing_sriov :: _ ->
>       let existing_phy_pif = Db.Network_sriov.get_physical_PIF ~__context ~self:existing_sriov in
> ```

You can get this from your pattern match if you replace the last `_`?

---

*[ocaml/xapi/xapi_vlan.ml, line 56 at r3](https://reviewable.io:443/reviews/xapi-project/xen-api/3479#-L6R3vYaX9MAnVDkGRyc:-L6R3vYaX9MAnVDkGRyd:b-9phv6n) ([raw file](https://github.com/xapi-project/xen-api/blob/d84ad666a28e89efffa83fd40e0bef9ee64caffb/ocaml/xapi/xapi_vlan.ml#L56)):*
> ```OCaml
>   begin
>     let open Xapi_pif_helpers in
>     match pif_topo with
> ```

Perhaps we can push this logic directly into the helpers, to keep this function plain and simple.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xen-api/3479)*

**Comment by krizex** [comment]
Review status: 9 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[ocaml/xapi/xapi_network_helpers.ml, line 41 at r3](https://beta.reviewable.io/reviews/xapi-project/xen-api/3479#-L6R3iybjTpWd2ei4T5D:-L6RfDLe1Orsqf9FzEaU:b-896fix) ([raw file](https://github.com/xapi-project/xen-api/blob/d84ad666a28e89efffa83fd40e0bef9ee64caffb/ocaml/xapi/xapi_network_helpers.ml#L41)):*
<details><summary><i>Previously, robhoes (Rob Hoes) wrote…</i></summary><blockquote>

There is another related function in xapi_network_sriov_helpers.ml: `assert_sriov_pif_compatible_with_network`.
</blockquote></details>

Done.

---

*[ocaml/xapi/xapi_network_helpers.ml, line 48 at r3](https://beta.reviewable.io/reviews/xapi-project/xen-api/3479#-L6R2dgwSBwZrZmaCQfw:-L6RfzaoAHhHmJX6CawE:burb44k) ([raw file](https://github.com/xapi-project/xen-api/blob/d84ad666a28e89efffa83fd40e0bef9ee64caffb/ocaml/xapi/xapi_network_helpers.ml#L48)):*
<details><summary><i>Previously, robhoes (Rob Hoes) wrote…</i></summary><blockquote>

You can get this from your pattern match if you replace the last `_`?
</blockquote></details>

No, I need a `pif_ref` but from the pattern match I only can get a `pif_rec`

---

*[ocaml/xapi/xapi_vlan.ml, line 56 at r3](https://beta.reviewable.io/reviews/xapi-project/xen-api/3479#-L6R3vYaX9MAnVDkGRyc:-L6RkDaP4U07dB7eP2rS:b-896fix) ([raw file](https://github.com/xapi-project/xen-api/blob/d84ad666a28e89efffa83fd40e0bef9ee64caffb/ocaml/xapi/xapi_vlan.ml#L56)):*
<details><summary><i>Previously, robhoes (Rob Hoes) wrote…</i></summary><blockquote>

Perhaps we can push this logic directly into the helpers, to keep this function plain and simple.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/xapi-project/xen-api/3479)*

**Comment by krizex** [comment]
Updated.

---

Review status: 0 of 13 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xen-api/3479#-:-L6RlpEydam8H4Q7m8Qo:b-bw8e14)*

**Comment by robhoes** [comment]
:lgtm:

---

Reviewed 13 of 13 files at r10, 4 of 4 files at r11, 7 of 7 files at r12, 3 of 3 files at r13, 3 of 3 files at r14.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/xapi-project/xen-api/3479#-:-L6RyN2CtGexse3lwqcj:bnfp4nl)*

**Pull request merged**